engine/resources/packagekit/packagekit.go

```diff
@@ -186,34 +187,50 @@ func (obj *Conn) Close() error {
 }
 
 // internal helper to add signal matches to the bus, should only be called once
-func (obj *Conn) matchSignal(ch chan *dbus.Signal, path dbus.ObjectPath, iface string, signals []string) error {
+func (obj *Conn) matchSignal(ch chan *dbus.Signal, path dbus.ObjectPath, iface string, signals []string) (func() error, error) {
 	if obj.Debug {
 		obj.Logf("matchSignal(%v, %v, %s, %v)", ch, path, iface, signals)
 	}
 	// eg: gdbus monitor --system --dest org.freedesktop.PackageKit --object-path /org/freedesktop/PackageKit | grep <signal>
+	bus := obj.GetBus().BusObject()
 	var call *dbus.Call
+	var argsList []string
+	// cleanup function should be called when done or when AddMatch errors
+	removeSignals := func() error {
+		var errList error
+		for i := len(argsList) - 1; i >= 0; i-- { // last in first out
+			if call = bus.Call(engineUtil.DBusRemoveMatch, 0, argsList[i]); call.Err != nil {
+				errList = multierr.Append(errList, call.Err)
+			}
+		}
+		return errList
+	}
 	// TODO: if we make this call many times, we seem to receive signals
 	// that many times... Maybe this should be an object singleton?
-	bus := obj.GetBus().BusObject()
 	pathStr := fmt.Sprintf("%s", path)
 	if len(signals) == 0 {
-		call = bus.Call(engineUtil.DBusAddMatch, 0, "type='signal',path='"+pathStr+"',interface='"+iface+"'")
+		args := fmt.Sprintf("type='signal', path='%s', interface='%s'", pathStr, iface)
+		argsList = append(argsList, args)
+		call = bus.Call(engineUtil.DBusAddMatch, 0, args)
 	} else {
 		for _, signal := range signals {
-			call = bus.Call(engineUtil.DBusAddMatch, 0, "type='signal',path='"+pathStr+"',interface='"+iface+"',member='"+signal+"'")
-			if call.Err != nil {
-				break
+			args := fmt.Sprintf("type='signal', path='%s', interface='%s', member'%s'", pathStr, iface, signal)
+			argsList = append(argsList, args)
+			if call = bus.Call(engineUtil.DBusAddMatch, 0, args); call.Err != nil {
+				break // fail if any one fails
 			}
 		}
 	}
 	if call.Err != nil {
-		return call.Err
+		defer removeSignals() // ignore the error
```

Comment:
Are you ok with ignoring the error here? i.e. if the 3rd of 4 fail, it will try to remove all 4 signals, even though only 2 were added in the first place. The last two will fail silently, since there's nothing to do.

Reply:
Yeah, the issue of "we're closing/shutting down" but on close there's an error doesn't really mean something good happened, but it's not something we can often do anything about, so in this case it's fine AFAICT.